File: Assets/Scripts/Components/Manager/GameManager.cs
using UnityEngine;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

/**
 * 
 */
public class GameManager {
	
	public int TetrisAmount = 5;												// Amount of Tetris in List
			
	protected Map _map;
	protected Tetris _currentTetris;
	protected List<Tetris> _tetrisList;
	protected Dictionary <String, LevelStrategyBase> _levelStrategyDictionary;
	protected List<GameRule> _gameRules;
	protected PlayerCredit _playerCredit;

	public delegate void TetrisEventHanlder (TetrisEvent e);
	public static event TetrisEventHanlder _eventCallback;
	
	public Map Map {
		get { 
			return _map;
		}
	}
	public Tetris CurrentTetris {
		get { 
			return _currentTetris;
		}
	}
	public List<Tetris> TetrisList {
		get { 
			return _tetrisList;
		}
	}
	public GameRule CurrentRule{
		get { 
			return _gameRules [_playerCredit.level];
		}
	}
	public PlayerCredit PlayerCredit {
		get { 
			return _playerCredit;
		}
	}
	
	
	float time = 0.0f;																//Current Time, used to controll Tetris falling
	int score = 0;																	//Player Score, used to Level up
	bool isSpeedup = false;													//Wether in "Speed Up" mode
	bool gameRuleInitCompleted = false;								//Wether Levels Infos were loaded

	
	/**
     *  Constructor
     */
	public GameManager () {
		_map = new Map ( 10, 20);
		_tetrisList = new List<Tetris> ();	
		_gameRules = new List<GameRule> ();
		_playerCredit = new PlayerCredit ();
		_levelStrategyDictionary = new Dictionary<String , LevelStrategyBase> ();
		_levelStrategyDictionary["simple"] = new LevelStrategy_Simple();
		_levelStrategyDictionary["normal"] = new LevelStrategy_Normal();
		_levelStrategyDictionary["advanced"] = new LevelStrategy_Advanced();
		
		TextAsset txt = Resources.Load ("GameRule",typeof(TextAsset)) as TextAsset;
		string [] lineArray = txt.text.Split ("\r\n"[0]);
		for (int i = 1;i < lineArray.Length-1;i++)
		{
			var ruleValue = lineArray[i].Split(","[0]);
//			Debug.Log (ruleValue[0]+":::::"+ruleValue[1]+":::::"+ruleValue[2]);
			GameRule rule = new GameRule (ruleValue[0],ruleValue[1],ruleValue[2]);
			_gameRules.Add (rule);
		}
		gameRuleInitCompleted = true;
	}
   
    public void step() {
		if(gameRuleInitCompleted==false)
			return;
		
		GameRule rule = _gameRules [_playerCredit.level];
		if (_currentTetris == null){
		
			// if 1st time, Generate Tetris List
			if (_tetrisList.Count == 0){
				for(int i=0;i<5;i++){
					LevelStrategyBase ls =(LevelStrategyBase)_levelStrategyDictionary[rule.LevelStrategy];
					_tetrisList.Add(ls.execute().create());
				}
			}
			
			// get CurrentTetris from List
			_currentTetris = _tetrisList[0];
			_tetrisList.RemoveAt(0);
			
			// Set up CurrentTetris
			float xPos = Mathf.Floor (((float) _map.Width - (float) _currentTetris.Width)/2);
			float yPos = (float) _map.Height;
			_currentTetris.Postion = new Vector2 (xPos,yPos);
			
			//add Another Tetris to List
			LevelStrategyBase s =(LevelStrategyBase)_levelStrategyDictionary[rule.LevelStrategy];
			_tetrisList.Add(s.execute().create());
		};
			
			
		time += Time.deltaTime;

		float speed = isSpeedup ? 1/(rule.Speed*5) : 1/rule.Speed ;

		// check the coliision of _currentTetris
		while (_currentTetris != null  && time > speed){
			_currentTetris.Postion = new Vector2 (_currentTetris.Postion.x,_currentTetris.Postion.y - 1);
			if (collisionDetection(_currentTetris)) {
				_currentTetris.Postion = new Vector2 (_currentTetris.Postion.x,_currentTetris.Postion.y + 1);
				putTetrisToMap (_currentTetris);
				updateElements();
				_currentTetris = null;
			}
			else{
				if ( _eventCallback != null){
					_eventCallback(TetrisEvent.CHANGE_POSITION);
				}
			}

			time -= speed;
		}

    }

    public void input(Operator op ) {
		if (_currentTetris == null )
			return;
		
		switch (op) {
		
			#region LEFT
		case Operator.LEFT:
			_currentTetris.Postion = new Vector2 (_currentTetris.Postion.x  - 1,_currentTetris.Postion.y);
			if ( collisionDetection(_currentTetris)){
				_currentTetris.Postion = new Vector2 (_currentTetris.Postion.x  + 1,_currentTetris.Postion.y);
			}
			else{
				if (_eventCallback != null){
					_eventCallback( TetrisEvent.CHANGE_POSITION);
				}
			}
			break;
			#endregion
		
			#region RIGHT
		case Operator.RIGHT:
			_currentTetris.Postion = new Vector2 (_currentTetris.Postion.x  + 1,_currentTetris.Postion.y);
			if ( collisionDetection(_currentTetris)){
				_currentTetris.Postion = new Vector2 (_currentTetris.Postion.x  - 1,_currentTetris.Postion.y);
			}
			else{
				if (_eventCallback != null){
					_eventCallback( TetrisEvent.CHANGE_POSITION);
				}
			}
			break;
			#endregion
			
			#region DIRECT_FALL
		case Operator.DIRECT_FALL:
			while (_currentTetris != null) {
				_currentTetris.Postion = new Vector2 (_currentTetris.Postion.x,_currentTetris.Postion.y - 1);
				if ( collisionDetection (_currentTetris)){
					_currentTetris.Postion = new Vector2 (_currentTetris.Postion.x,_currentTetris.Postion.y + 1);
					putTetrisToMap (_currentTetris);
					_currentTetris = null;
					if (_eventCallback != null){
						_eventCallback( TetrisEvent.CHANGE_POSITION);
					}
					updateElements();
				}
			}
			break;
			#endregion
			
			#region SPEEDUP_START
		case Operator.SPEEDUP_START:
			isSpeedup = true;
			if (_eventCallback != null){
				_eventCallback( TetrisEvent.CHANGE_POSITION);
			}
			break;
			#endregion
			
			#region SPEEDUP_END
		case Operator.SPEEDUP_END:
			isSpeedup = false;
			if (_eventCallback != null){
				_eventCallback( TetrisEvent.CHANGE_POSITION);
			}
			break;
			#endregion
		
			#region TURN
		case Operator.TURN:
			Tetris tmpTetris = _currentTetris.cloneTetris();
			_currentTetris = rotateAndShiftTetris (tmpTetris);
			if (_eventCallback != null){
				_eventCallback( TetrisEvent.CHANGE_POSITION);
			}
			break;
			#endregion
		}
    }

	protected bool collisionDetection(Tetris tetris) {

		for(int x=0;x<tetris.Width;x++){
			for(int y=0;y<tetris.Height;y++){
				if (tetris.Shape[x,y].isNull == false){
					//Check bottom boundary
					if (tetris.Postion.y + y < 0) {
						return true;
					}
					//Check left boundary
					if (tetris.Postion.x + x < 0) {
						return true;
					}
					//Check right boudary
					if (tetris.Postion.x + x >= _map.Width) {
						return true;
					}
					// Check Map Elements
					if (tetris.Postion.y + y < _map.Height) {
						Element e = _map.Elements [(int)tetris.Postion.y + y][(int)tetris.Postion.x + x];
						if (e != null && e.isNull == false) {
							return true;
						}
					}
				}	
			}
		}
        return false;
    }
        
	protected void putTetrisToMap(Tetris tetris) {
		for(int x=0;x<tetris.Width;x++){
			for(int y=0;y<tetris.Height;y++){
				if (_currentTetris.Shape[x,y].isNull == false){
					//check GameOver
					if (_currentTetris.Postion.y + y >= _map.Height){
						if (_eventCallback != null ) {
							_eventCallback (TetrisEvent.GAME_OVER);
						}
						// Restart game
						_map.clearElements();
						_playerCredit.level = 0;
						_playerCredit.score = 0;
						_tetrisList = null ;
						_tetrisList = new List<Tetris> ();	
					}
					//put Tetris to Map
					else {
						_map.Elements [(int)_currentTetris.Postion.y + y][(int)_currentTetris.Postion.x + x].isNull = false;
					}			
					
				}
			}
		}
	}
	
	protected void updateElements () {
		//checkElements
		int clearCount = _map.checkElements();
		if (clearCount != 0 )
		{
			if (_eventCallback != null ) {
				_eventCallback (TetrisEvent.ATTACH);
			}
		}
		
		//update Player Score
		_playerCredit.score += clearCount * clearCount * 10;
		
		//update Player Level
		GameRule rule = _gameRules [_playerCredit.level];
		if (_playerCredit.score - score > rule.Score) {
			score +=  rule.Score;
			_playerCredit.level ++ ;
		}
		
	}
	
	protected Tetris rotateAndShiftTetris (Tetris tmp) {
		Tetris tmpTetris = tmp;
		//Rotate tmpTetris
		tmpTetris.rotateTetris();
		//Shift tmpTetris in 1 space
		tmpTetris = shiftTetris (tmpTetris, 1);
		return tmpTetris;
	}
	
    
	protected Tetris shiftTetris (Tetris tmpTetris, int shiftCount)
	{
		//Shift Order: Down > Up > Left > Right
		if (collisionDetection (tmpTetris)) {
			//Shift Down
			Debug.Log ("Shift Down" + shiftCount );
			tmpTetris.Postion = new Vector2 (tmpTetris.Postion.x, tmpTetris.Postion.y - shiftCount);
			if (collisionDetection (tmpTetris)) {
				tmpTetris.Postion = new Vector2 (tmpTetris.Postion.x, tmpTetris.Postion.y + shiftCount);
				//Shift Up
				Debug.Log ("Shift Up"+ shiftCount );
				tmpTetris.Postion = new Vector2 (tmpTetris.Postion.x, tmpTetris.Postion.y + shiftCount);
				if (collisionDetection (tmpTetris)) {
					tmpTetris.Postion = new Vector2 (tmpTetris.Postion.x, tmpTetris.Postion.y - shiftCount);
					//Shift Left
					Debug.Log ("Shift Left"+ shiftCount );
					tmpTetris.Postion = new Vector2 (tmpTetris.Postion.x - shiftCount, tmpTetris.Postion.y);
					if (collisionDetection (tmpTetris)) {
						tmpTetris.Postion = new Vector2 (tmpTetris.Postion.x + shiftCount, tmpTetris.Postion.y);
						//Shift Right
						Debug.Log ("Shift Right"+ shiftCount );
						tmpTetris.Postion = new Vector2 (tmpTetris.Postion.x + shiftCount, tmpTetris.Postion.y);
						if (collisionDetection (tmpTetris)) {
							tmpTetris.Postion = new Vector2 (tmpTetris.Postion.x - shiftCount, tmpTetris.Postion.y);
							Debug.Log ("TetrisRoate: Can not Shift in " + shiftCount +" space");
							return shiftTetris (tmpTetris,shiftCount+1); //Shift tmpTetris in +1 space
						}
					}
				}
			}
		}
		return tmpTetris;
	}
}

File: src/src/manager/GameManager.js
/**
 * Created by JasonWu on 6/17/15.
 */
var GameManager = cc.Class.extend({

    map:null,//用于保存当前地图的信息

    currentTetris:null,//当前的砖头

    tetrisList:null,//当前剩余的砖头

    levelStrategyHashMap:null,//用于存放游戏策略的Map

    gameRule:null,//当前所有关卡的信息

    playerCredit:null,//玩家当前的信息

    eventCallback:null,//事件回调函数

    _gameRuleInitCompleted:false,//表示关卡信息是否加载玩过

    _time:0,//当前执行的时间 每帧增加一点，用于控制砖头移动

    _isSpeedup:false,//当前是否加速状态

    _score:0,//当前累计所要达到的积分

    ctor:function(){
        this.map=new Map(10,20);
        this.tetrisList=[];
        this.playerCredit=new PlayerCredit();
        this.gameRule=[];
        this.levelStrategyHashMap={};
        this.levelStrategyHashMap["simple"]=new SimpleLevelStrategy();
        this.levelStrategyHashMap["normal"]=new NormalLevelStrategy();
        this.levelStrategyHashMap["advanced"]=new AdvancedLevelStrategy();

        //异步加载资源
        var me=this;
        cc.loader.loadTxt(res.GameRule_csv,function(err,txt){

            var ruleStrList=txt.split("\r\n");

            for(var i=1;i<ruleStrList.length;i++)
            {
                var ruleValue=ruleStrList[i].split(",");
                var rule=new GameRule();
                rule.speed=parseInt(ruleValue[0]);
                rule.score=parseInt(ruleValue[1]);
                rule.strategy=ruleValue[2];
                me.gameRule.push(rule);
            }

            me._gameRuleInitCompleted=true;

        });



    },

    step:function(){
        if(this._gameRuleInitCompleted==false)
            return;


        this._time+=cc.director.getDeltaTime();

        var rule=this.gameRule[this.playerCredit.level];

        //获取砖头
        if(this.currentTetris==null)
        {
            //如何是第一次进入，则创建一定数量的砖头
            if(this.tetrisList.length==0){
                for(var i=0;i<5;i++){
                    var s=this.levelStrategyHashMap[rule.strategy];
                    this.tetrisList.push(s.execute().create());
                }
            }

            //取出砖头
            this.currentTetris=this.tetrisList.shift();
            this.currentTetris.position.x=Math.floor((this.map.width-this.currentTetris.width)/2);
            this.currentTetris.position.y=this.map.height;

            //再存入新的
            var s=this.levelStrategyHashMap[rule.strategy];
            this.tetrisList.push(s.execute().create());

        }

        //向下移动砖头


        var speed=1/rule.speed;//计算出每格所需的时间  如果时间满足则可以移动
        if(this._isSpeedup)speed/=3;//加速


        while(this.currentTetris!=null && this._time>speed)
        {
            this.currentTetris.position.y-=1;

            //检查碰撞
            if(this.collisionDetection(this.currentTetris)) {
                //吸附砖头 抛出事件
                this.currentTetris.position.y += 1;
                this.putTetrisToMap(this.currentTetris);

                this.checkElements();

                this.currentTetris = null;

            }else{
                if (this.eventCallback != null)
                    this.eventCallback(TetrisEvent.CHANGE_POSITION);

            }

            this._time-=speed;
        }





    },

    input:function(operator){
        if(this.currentTetris==null)
            return;

        switch(operator){
            case Operator.LEFT:


                this.currentTetris.position.x--;

                if(this.collisionDetection(this.currentTetris))
                {
                    this.currentTetris.position.x++;
                }else{
                    //抛出事件
                    if (this.eventCallback != null)
                        this.eventCallback(TetrisEvent.CHANGE_POSITION);
                }

                break;
            case Operator.RIGHT:

                this.currentTetris.position.x++;

                if(this.collisionDetection(this.currentTetris))
                {
                    this.currentTetris.position.x--;
                }else{
                    //抛出事件
                    if (this.eventCallback != null)
                        this.eventCallback(TetrisEvent.CHANGE_POSITION);
                }

                break;
            case Operator.DERECT_FALL:
                //直接落地

                while(this.currentTetris!=null)
                {
                    this.currentTetris.position.y-=1;

                    if(this.collisionDetection(this.currentTetris))
                    {
                        this.currentTetris.position.y+=1;

                        this.putTetrisToMap(this.currentTetris);

                        this.checkElements();

                        this.currentTetris = null;
                    }
                }

                break;
            case Operator.SPEEDUP_START:
                //开始加速
                this._isSpeedup=true;
                break;
            case Operator.SPEEDUP_END:
                //结束加速
                this._isSpeedup=false;
                break;
            case Operator.TURN:
                //旋转

                var tmpTetris= this.currentTetris.clone();
                tmpTetris.rotate();

                var me=this;
                var testList=[];
                testList.push(function(t){
                    //检测碰撞
                    return me.collisionDetection(t);
                });
                testList.push(function(t){
                    //向下移动 进行检测
                    t.position.y-=1;
                    var result= me.collisionDetection(t);
                    if(result)
                    {
                        t.position.y+=1;
                    }
                    return result;
                });
                testList.push(function(t){
                    //向上检测
                    t.position.y+=1;
                    var result= me.collisionDetection(t);
                    if(result)
                    {
                        t.position.y-=1;
                    }
                    return result;
                });
                testList.push(function(t){
                    //向左检测
                    t.position.x-=1;
                    var result= me.collisionDetection(t);
                    if(result)
                    {
                        t.position.x+=1;
                    }
                    return result;
                });
                testList.push(function(t){
                    //向右检测
                    t.position.x+=1;
                    var result= me.collisionDetection(t);
                    if(result)
                    {
                        t.position.x-=1;
                    }
                    return result;
                });

                var isCancel=true;
                for(var i=0;i<testList.length;i++)
                {
                    if(testList[i](tmpTetris)==false)
                    {
                        isCancel=false;
                        break;
                    }
                }

                this.currentTetris=tmpTetris;

                break;
        }

    },

    collisionDetection:function(tetris){

        for(var x=0;x<tetris.width;x++)
        {
            for(var y=0;y<tetris.height;y++)
            {
                var e=tetris.shape[x][y];
                if(e.isNull==false)
                {
                    //检测地图 注意：地图数据是先行再列[y][x]

                    if(y+ tetris.position.y<0)//接触到底边
                        return true;

                    if(x+ tetris.position.x<0)//接触到左边
                        return true;

                    if(x+tetris.position.x>=this.map.width)//接触到右边
                        return true;

                    var row=this.map.map[y+ tetris.position.y];
                    if(row==null)
                        continue;
                    var e=row[x+ tetris.position.x];
                    if(e==null)
                        continue;
                    if(e.isNull==false)
                    {
                        return true;
                    }
                }
            }
        }
        return false;
    },

    putTetrisToMap:function(tetris){

        for(var x=0;x<tetris.width;x++)
        {
            for(var y=0;y<tetris.height;y++)
            {
                var e=tetris.shape[x][y];
                if(e.isNull==false)
                {
                    if(y+ tetris.position.y>=this.map.height)
                    {
                        //游戏结束
                        if (this.eventCallback != null)
                            this.eventCallback(TetrisEvent.GAME_OVER);

                        return;
                    }

                    //填充数据 注意：地图数据是先行再列[y][x]
                    this.map.map[y+ tetris.position.y][x+ tetris.position.x].isNull=false;
                }
            }
        }

    },

    checkElements:function(){
        //检查消除

        var rule=this.gameRule[this.playerCredit.level];

        var clearCount = this.map.checkElements();
        if (clearCount != 0) {
            //加分
            this.playerCredit.score += 10 * clearCount * clearCount;
            //超过一定的关卡积分则升级
            if (this.playerCredit.score > this._score+rule.score) {
                this._score+=rule.score;
                this.playerCredit.level++;
            }
            //抛出事件
            if (this.eventCallback != null)
                this.eventCallback(TetrisEvent.ATTACH);
        }

    }

});